File: 07.compiler/divider.go
package main

func NewDivider() *Divider {
	return &Divider{}
}

func (d *Divider) DivideClassVarDecs(content []Tokenized) [][]Tokenized {
    var classVarDecs [][]Tokenized
    startIndex := 0

    for startIndex < len(content) {
        keywordIndex := -1
        for i := startIndex; i < len(content); i++ {
            if contains(classVarDec, content[i].Value) {
                keywordIndex = i
                break
            }
        }

        if keywordIndex == -1 {
            break
        }

        semicolonIndex := -1
        for i := keywordIndex + 1; i < len(content); i++ {
            if content[i].Value == ";" {
                semicolonIndex = i
                break
            }
        }

        if semicolonIndex == -1 {
            break
        }

        declaration := content[keywordIndex : semicolonIndex+1]
        classVarDecs = append(classVarDecs, declaration)

        startIndex = semicolonIndex
    }

    return classVarDecs
}

func (d *Divider) DivideSubroutines(content []Tokenized) [][]Tokenized {
    var subroutines [][]Tokenized
    startIndex := 0

    for startIndex < len(content) {
        keywordIndex := -1
        for i := startIndex; i < len(content); i++ {
            if contains(subroutineDec, content[i].Value) {
                keywordIndex = i
                break
            }
        }

        if keywordIndex == -1 {
            break
        }

        closingBracketIndex := -1
        bracketCount := 0
        for i := keywordIndex; i < len(content); i++ {
            if content[i].Value == "{" {
                bracketCount++
            } else if content[i].Value == "}" {
                bracketCount--
                if bracketCount == 0 {
                    closingBracketIndex = i
                    break
                }
            }
        }

        if closingBracketIndex == -1 {
            break
        }

        subroutine := content[keywordIndex : closingBracketIndex+1]
        subroutines = append(subroutines, subroutine)
        startIndex = closingBracketIndex
    }

    return subroutines
}

func (d *Divider) DivideParameterList(content []Tokenized) [][]Tokenized {
	openingBracketIndex := -1
	for i, token := range content {
		if token.Value == "(" {
			openingBracketIndex = i
			break
		}
	}

	closingBracketIndex := -1
	for i := openingBracketIndex + 1; i < len(content); i++ {
		if content[i].Value == ")" && i+1 < len(content) && content[i+1].Value == "{" {
			closingBracketIndex = i
			break
		}
	}

	if openingBracketIndex == -1 || closingBracketIndex == -1 {
		return nil
	}

	parameters := content[openingBracketIndex+1 : closingBracketIndex]
	var result [][]Tokenized
	start := 0

	for start < len(parameters) {
		commaIndex := -1
		for i := start; i < len(parameters); i++ {
			if parameters[i].Value == "," {
				commaIndex = i
				break
			}
		}

		if commaIndex != -1 {
			result = append(result, parameters[start:commaIndex])
			start = commaIndex + 1
		} else {
			result = append(result, parameters[start:])
			break
		}
	}

	return result
}

func (d *Divider) DivideSubroutineBody(content []Tokenized) []Tokenized {
    openingBracketIndex := -1
    for i := 0; i < len(content); i++ {
        if i > 0 && content[i].Value == "{" && content[i-1].Value == ")" {
            openingBracketIndex = i
            break
        }
    }

    return content[openingBracketIndex+1 : len(content)-1]
}

func (d *Divider) DivideVarDecs(content []Tokenized) [][]Tokenized {
    var varDecs [][]Tokenized
    startIndex := 0

    for startIndex < len(content) {
        keywordIndex := -1
        for i := startIndex; i < len(content); i++ {
            if content[i].Value == "var" {
                keywordIndex = i
                break
            }
        }

        if keywordIndex == -1 {
            break
        }

        semicolonIndex := -1
        for i := keywordIndex + 1; i < len(content); i++ {
            if content[i].Value == ";" {
                semicolonIndex = i
                break
            }
        }

        if semicolonIndex == -1 {
            break
        }

        declaration := content[keywordIndex : semicolonIndex+1]
        varDecs = append(varDecs, declaration)

        startIndex = semicolonIndex
    }

    return varDecs
}

func (d *Divider) DivideStatements(content []Tokenized) [][]Tokenized {
    var result [][]Tokenized
    for i := 0; i < len(content); i++ {
        if contains(statements, content[i].Value) {
            typ := content[i].Value.(string)

            switch typ {
            case "let", "do", "return":
                closingBracketIndex := -1
                for j := i + 1; j < len(content); j++ {
                    if content[j].Value == ";" {
                        closingBracketIndex = j
                        break
                    }
                }

                if closingBracketIndex == -1 {
                    break
                }

                result = append(result, content[i:closingBracketIndex+1])
                i = closingBracketIndex

            case "while":
                openBracketCount := 0
                closingBracketIndex := -1
                for j := i; j < len(content); j++ {
                    if content[j].Value == "{" {
                        openBracketCount++
                    } else if content[j].Value == "}" {
                        openBracketCount--
                        if openBracketCount == 0 {
                            closingBracketIndex = j
                            break
                        }
                    }
                }

                result = append(result, content[i:closingBracketIndex+1])
                i = closingBracketIndex

            case "if":
                openBracketCount := 0
                closingBracketIndex := -1
                for j := i; j < len(content); j++ {
                    if content[j].Value == "{" {
                        openBracketCount++
                    } else if content[j].Value == "}" {
                        openBracketCount--
                        if openBracketCount == 0 {
                            hasElse := j+1 < len(content) && content[j+1].Value == "else"
                            if !hasElse || j+1 == len(content) {
                                closingBracketIndex = j
                                break
                            }
                        }
                    }
                }

                result = append(result, content[i:closingBracketIndex+1])
                i = closingBracketIndex
            }
        }
    }

    return result
}

func (d *Divider) DivideExpression(content []Tokenized) []Tokenized {
	openBracket := -1
	for i, item := range content {
		if item.Value == "(" {
			openBracket = i
			break
		}
	}
 
	closingBracket := -1
	openBracketCount := 0
	for i := openBracket; i < len(content); i++ {
		if content[i].Value == "(" {
			openBracketCount++
		} else if content[i].Value == ")" {
			openBracketCount--
			if openBracketCount == 0 {
				closingBracket = i
				break
			}
		}
	}
 
	return content[openBracket+1 : closingBracket]
}

type ifStatements struct {
    If    [][]Tokenized
    Else  [][]Tokenized
}

func (d *Divider) DivideIfStatements(content []Tokenized) ifStatements {
    result := ifStatements{
        If:   [][]Tokenized{},
        Else: [][]Tokenized{},
    }

    openBracket := -1
    for i, item := range content {
        if item.Value == "{" {
            openBracket = i
            break
        }
    }

    openBracketCount := 0
    closingBracket := -1
    for i := openBracket; i < len(content); i++ {
        if content[i].Value == "{" {
            openBracketCount++
        }
        if content[i].Value == "}" {
            openBracketCount--
            if openBracketCount == 0 {
                closingBracket = i
                break
            }
        }
    }

    ifStatements := d.DivideStatements(content[openBracket+1 : closingBracket])
    result.If = ifStatements

    elseIndex := -1
    openElseBracketCount := 0
    for i := closingBracket + 1; i < len(content); i++ {
        if content[i].Value == "{" {
            openElseBracketCount++
        }

        if content[i].Value == "}" {
            openElseBracketCount--
        }

        if content[i].Value == "else" && openElseBracketCount == 0 {
            elseIndex = i
            break
        }
    }

    if elseIndex != -1 {
        openBracketCount := 0
        openBracketElse := -1
        for i := elseIndex; i < len(content); i++ {
            if content[i].Value == "{" {
                openBracketElse = i
                break
            }
        }

        closingBracketElse := -1
        for i := elseIndex; i < len(content); i++ {
            if content[i].Value == "{" {
                openBracketCount++
            }
            if content[i].Value == "}" {
                openBracketCount--
                if openBracketCount == 0 {
                    closingBracketElse = i
                    break
                }
            }
        }

        elseStatements := d.DivideStatements(content[openBracketElse+1 : closingBracketElse])
        result.Else = elseStatements
    }

    return result
}

func (d *Divider) DivideWhileStatements(content []Tokenized) [][]Tokenized {
    openBracket := -1
    for i, item := range content {
        if item.Value == "{" {
            openBracket = i
            break
        }
    }

    closingBracket := -1
    openBracketCount := 0
    for i := openBracket; i < len(content); i++ {
        if content[i].Value == "{" {
            openBracketCount++
        } else if content[i].Value == "}" {
            openBracketCount--
            if openBracketCount == 0 {
                closingBracket = i
                break
            }
        }
    }

    whileStatements := d.DivideStatements(content[openBracket+1 : closingBracket])
    return whileStatements
}

func (d *Divider) DivideTerms(content []Tokenized) [][]Tokenized {
    var result [][]Tokenized
    
    if containsDot(content) {
        result = append(result, content)
        return result
    }

    for i := 0; i < len(content); i++ {
        token := content[i]
        
        if contains(operands, token.Value) || contains(unaryOperands, token.Value) {
            result = append(result, []Tokenized{token})
            continue
        }

        if token.Value == "(" {
            closingBracket := findClosingBracket(content, i, "(", ")")
            result = append(result, content[i:closingBracket+1])
            i = closingBracket
            continue
        } else if token.Value == "[" {
            closingBracket := findClosingBracket(content, i, "[", "]")
            result = append(result, content[i:closingBracket+1])
            i = closingBracket
            continue
        }

        result = append(result, []Tokenized{token})
    }

    return result
}

func (d *Divider) DivideVarTerms(content []Tokenized) [][]Tokenized {
    var result [][]Tokenized
    
    if !containsOperands(content) {
        return [][]Tokenized{content}
    }

    for i := 0; i < len(content); i++ {
        token := content[i]

        if contains(operands, token.Value) {
            result = append(result, []Tokenized{token})

            if i+1 < len(content) && content[i+1].Value == "(" {
                closingBracket := findClosingBracket(content, i+1, "(", ")")
                result = append(result, content[i+1:closingBracket+1])
                i = closingBracket
            }
        } else if token.Value == "(" {
            closingBracket := findClosingBracket(content, i, "(", ")")
            result = append(result, content[i:closingBracket+1])
            i = closingBracket
        } else if token.Value == "[" {
            closingBracket := findClosingBracket(content, i, "[", "]")
            result = append(result, content[i:closingBracket+1])
            i = closingBracket
        } else {
            result = append(result, []Tokenized{token})
        }
    }

    return result
}

func (d *Divider) DivideExpressionList(content []Tokenized) [][]Tokenized {
    var result [][]Tokenized
    
    if !containsComma(content) && len(content) > 0 {
        result = append(result, content)
        return result
    }

    for i := 0; i < len(content); i++ {
        commaIndex := -1
        for j := i; j < len(content); j++ {
            if content[j].Value == "," {
                commaIndex = j
                break
            }
        }

        if commaIndex != -1 {
            result = append(result, content[i:commaIndex])
            result = append(result, []Tokenized{content[commaIndex]})
            i = commaIndex
        } else {
            if i != len(content) {
                result = append(result, content[i:])
                break
            }
        }
    }

    return result
}

func (d *Divider) ShuffleTerms(content [][]Tokenized) [][]Tokenized {
    var operandIndexes []int
    
    for i, item := range content {
        if contains(operands, item[0].Value) || contains(unaryOperands, item[0].Value) {
            operandIndexes = append(operandIndexes, i)
        }
    }

    if len(operandIndexes) != 0 {
        for _, index := range operandIndexes {
            operand := content[index]
            right := content[index+1]
            content[index] = right
            content[index+1] = operand
        }
    }

    return content
}

File: draft/compiler/divider.ts
import {
    classVarDec,
    operands,
    statements,
    subroutineDec,
    unaryOperands,
} from './jack-grammar';
import { IDivider, Tokenized, ifStatements } from './types';

export class Divider implements IDivider {
    divideClassVarDecs(content: Tokenized[]): Tokenized[][] {
        const classVarDecs: Tokenized[][] = [];
        let startIndex = 0;

        while (startIndex < content.length) {
            let keywordIndex = content.findIndex((item, index) => {
                return (
                    classVarDec.includes(item.value as string) &&
                    index >= startIndex
                );
            });

            if (keywordIndex === -1) {
                break;
            }

            let semicolonIndex = content.findIndex(
                (item, index) => item.value === ';' && index > keywordIndex
            );

            if (semicolonIndex === -1) {
                break;
            }

            const declaration = content.slice(keywordIndex, semicolonIndex + 1);

            classVarDecs.push(declaration);

            startIndex = semicolonIndex;
        }

        return classVarDecs;
    }

    divideSubroutines(content: Tokenized[]): Tokenized[][] {
        const subroutines: Tokenized[][] = [];
        let startIndex = 0;

        while (startIndex < content.length) {
            let keywordIndex = content.findIndex((item, index) => {
                return (
                    subroutineDec.includes(item.value as string) &&
                    index >= startIndex
                );
            });

            if (keywordIndex === -1) {
                break;
            }

            let closingBracketIndex = -1;
            let bracketCount = 0;
            outerLoop: for (let i = keywordIndex; i < content.length; i++) {
                if (content[i].value === '{') {
                    bracketCount++;
                } else if (content[i].value === '}') {
                    bracketCount--;
                    if (bracketCount === 0) {
                        closingBracketIndex = i;
                        break outerLoop;
                    }
                }
            }

            if (closingBracketIndex === -1) {
                break;
            }

            const subroutine = content.slice(
                keywordIndex,
                closingBracketIndex + 1
            );

            subroutines.push(subroutine);
            startIndex = closingBracketIndex;
        }

        return subroutines;
    }

    divideParameterList(content: Tokenized[]): Tokenized[][] {
        let openingBracketIndex = content.findIndex(
            (item) => item.value === '('
        );
        let closingBracketIndex = content.findIndex(
            (item, index) =>
                item.value === ')' && content[index + 1].value === '{'
        );

        const parameters = content.slice(
            openingBracketIndex + 1,
            closingBracketIndex
        );

        const list: Tokenized[][] = [];
        for (let i = 0; i < parameters.length; i++) {
            const comma = parameters.findIndex(
                (item, index) => item.value === ',' && index >= i
            );
            if (comma !== -1) {
                list.push(parameters.slice(i, comma));
                i = comma;
            } else {
                list.push(parameters.slice(i, parameters.length));
                break;
            }
        }
        return list;
    }

    divideSubroutineBody(content: Tokenized[]): Tokenized[] {
        let openingBracketIndex = content.findIndex(
            (item, index) =>
                item.value === '{' && content[index - 1].value === ')'
        );

        return content.slice(openingBracketIndex + 1, content.length - 1);
    }

    divideVarDecs(content: Tokenized[]): Tokenized[][] {
        const varDecs: Tokenized[][] = [];
        let startIndex = 0;

        while (startIndex < content.length) {
            let keywordIndex = content.findIndex((item, index) => {
                return item.value === 'var' && index >= startIndex;
            });

            if (keywordIndex === -1) {
                break;
            }

            let semicolonIndex = content.findIndex(
                (item, index) => item.value === ';' && index > keywordIndex
            );

            if (semicolonIndex === -1) {
                break;
            }

            const declaration = content.slice(keywordIndex, semicolonIndex + 1);
            varDecs.push(declaration);

            startIndex = semicolonIndex;
        }

        return varDecs;
    }

    divideStatements(content: Tokenized[]): Tokenized[][] {
        const result: Tokenized[][] = [];
        for (let i = 0; i < content.length; i++) {
            if (statements.includes(content[i].value as string)) {
                const type = content[i].value as string;

                switch (type) {
                    case 'let':
                    case 'do':
                    case 'return': {
                        let closingBracketIndex = content.findIndex(
                            (item, index) => item.value === ';' && index > i
                        );

                        if (closingBracketIndex === -1) {
                            break;
                        }

                        result.push(content.slice(i, closingBracketIndex + 1));
                        i = closingBracketIndex;
                        break;
                    }
                    case 'while': {
                        let openBracketCount = 0;
                        let closingBracketIndex = -1;
                        outerLoop: for (let j = i; j < content.length; j++) {
                            if (content[j].value === '{') {
                                openBracketCount++;
                            } else if (content[j].value === '}') {
                                openBracketCount--;
                                if (
                                    openBracketCount === 0 &&
                                    content[j + 1] !== undefined &&
                                    content[j + 1].value !== 'else'
                                ) {
                                    closingBracketIndex = j;
                                    break outerLoop;
                                } else if (openBracketCount === 0) {
                                    closingBracketIndex = j;
                                    break outerLoop;
                                }
                            }
                        }

                        result.push(content.slice(i, closingBracketIndex + 1));
                        i = closingBracketIndex;
                        break;
                    }
                    case 'if': {
                        let openBracketCount = 0;
                        let closingBracketIndex = -1;
                        outerLoop: for (let j = i; j < content.length; j++) {
                            if (content[j].value === '{') {
                                openBracketCount++;
                            } else if (content[j].value === '}') {
                                openBracketCount--;
                                if (
                                    openBracketCount === 0 &&
                                    content[j + 1] !== undefined &&
                                    content[j + 1].value !== 'else'
                                ) {
                                    closingBracketIndex = j;
                                    break outerLoop;
                                } else if (
                                    openBracketCount === 0 &&
                                    content[j + 1] == undefined
                                ) {
                                    closingBracketIndex = j;
                                    break outerLoop;
                                }
                            }
                        }
                        result.push(content.slice(i, closingBracketIndex + 1));
                        i = closingBracketIndex;
                        break;
                    }
                }
            }
        }

        return result;
    }

    divideExpression(content: Tokenized[]): Tokenized[] {
        const openBracket = content.findIndex((item) => item.value === '(');
        let closingBracket = -1;
        let openBracketCount = 0;
        for (let i = openBracket; i < content.length; i++) {
            if (content[i].value === '(') {
                openBracketCount++;
            } else if (content[i].value === ')') {
                openBracketCount--;
                if (openBracketCount === 0) {
                    closingBracket = i;
                    break;
                }
            }
        }
        return content.slice(openBracket + 1, closingBracket);
    }

    divideIfStatements(content: Tokenized[]): ifStatements {
        const result: ifStatements = {
            if: [],
            else: [],
        };

        let openBracketCount = 0;
        let closingBracket = -1;
        const openBracket = content.findIndex((item) => item.value === '{');
        for (let i = 0; i < content.length; i++) {
            if (content[i].value === '{') {
                openBracketCount++;
            }
            if (content[i].value === '}') {
                openBracketCount--;
                if (openBracketCount === 0) {
                    closingBracket = i;
                    break;
                }
            }
        }

        const ifStaments = this.divideStatements(
            content.slice(openBracket + 1, closingBracket)
        );
        result.if = ifStaments;

        let elseIndex = -1;
        let openElseBracketCount = 0;
        for (let i = closingBracket + 1; i < content.length; i++) {
            if (content[i].value === '{') {
                openElseBracketCount++;
            }

            if (content[i].value === '}') {
                openElseBracketCount--;
            }

            if (content[i].value === 'else' && openElseBracketCount === 0) {
                elseIndex = i;
                break;
            }
        }

        if (elseIndex !== -1) {
            let openBracketCount = 0;
            const openBracketElse = content.findIndex((item, index) => {
                return item.value === '{' && index > elseIndex;
            });
            let closingBracketElse = -1;
            for (let i = elseIndex; i < content.length; i++) {
                if (content[i].value === '{') {
                    openBracketCount++;
                }
                if (content[i].value === '}') {
                    openBracketCount--;
                    if (openBracketCount === 0) {
                        closingBracketElse = i;
                        break;
                    }
                }
            }
            const elseStatements = this.divideStatements(
                content.slice(openBracketElse + 1, closingBracketElse)
            );
            result.else = elseStatements;
        }
        return result;
    }

    divideWhileStatements(content: Tokenized[]): Tokenized[][] {
        const openBracket = content.findIndex((item) => item.value === '{');
        let closingBracket = -1;
        let openBracketCount = 0;
        for (let i = openBracket; i < content.length; i++) {
            if (content[i].value === '{') {
                openBracketCount++;
            } else if (content[i].value === '}') {
                openBracketCount--;
                if (openBracketCount === 0) {
                    closingBracket = i;
                    break;
                }
            }
        }

        const whileStatements = this.divideStatements(
            content.slice(openBracket + 1, closingBracket)
        );

        return whileStatements;
    }

    divideTerms(content: Tokenized[]): Tokenized[][] {
        const result: Tokenized[][] = [];
        if (content.some((item) => item.value === '.')) {
            result.push(content);
            return result;
        }
        // else if (
        //     content.some((item) => item.value === '[') &&
        //     content.length === 4
        // ) {
        //     result.push(content);
        //     return result;
        // }

        for (let i = 0; i < content.length; i++) {
            const token = content[i];
            if (
                operands.includes(token.value as string) ||
                unaryOperands.includes(token.value as string)
            ) {
                const token = content[i];
                result.push([token]);
                continue;
            }

            if (token.value === '(') {
                let closingBracket = -1;
                let openBracketCount = 0;

                for (let j = i; j < content.length; j++) {
                    if (content[j].value === '(') {
                        openBracketCount++;
                    } else if (content[j] && content[j].value === ')') {
                        openBracketCount--;
                        if (openBracketCount === 0) {
                            closingBracket = j;
                            break;
                        }
                    }
                }
                result.push(content.slice(i, closingBracket + 1));
                i = closingBracket;
                continue;
            } else if (token.value === '[') {
                let closingBracket = -1;
                let openBracketCount = 0;

                for (let j = i; j < content.length; j++) {
                    if (content[j].value === '[') {
                        openBracketCount++;
                    } else if (content[j].value === ']') {
                        openBracketCount--;
                        if (openBracketCount === 0) {
                            closingBracket = j;
                            break;
                        }
                    }
                }
                result.push(content.slice(i, closingBracket + 1));
                i = closingBracket;
                continue;
            }

            result.push([token]);
        }

        return result;
    }

    divideVarTerms(content: Tokenized[]): Tokenized[][] {
        const result: Tokenized[][] = [];
        if (!content.some((item) => operands.includes(item.value as string))) {
            return [content];
        }
        for (let i = 0; i < content.length; i++) {
            const token = content[i];

            if (operands.includes(token.value as string)) {
                result.push([token]);

                if (content[i + 1].value === '(') {
                    let closingBracket = -1;
                    let openBracketCount = 0;

                    for (let j = i + 1; j < content.length; j++) {
                        if (content[j].value === '(') {
                            openBracketCount++;
                        } else if (content[j].value === ')') {
                            openBracketCount--;
                            if (openBracketCount === 0) {
                                closingBracket = j;
                                break;
                            }
                        }
                    }

                    result.push(content.slice(i + 1, closingBracket + 1));
                    i = closingBracket;
                }

                // const nextOperand = content.findIndex(
                //     (item, index) =>
                //         operands.includes(item.value as string) && index > i
                // );
                // if (nextOperand === -1) {
                //     if (i !== content.length - 1) {
                //         result.push(content.slice(i + 1, content.length));
                //     }
                //     break;
                // } else {
                //     result.push(content.slice(i + 1, nextOperand));
                //     i = nextOperand;
                // }
            } else if (token.value === '(') {
                let closingBracket = -1;
                let openBracketCount = 0;

                for (let j = i; j < content.length; j++) {
                    if (content[j].value === '(') {
                        openBracketCount++;
                    } else if (content[j].value === ')') {
                        openBracketCount--;
                        if (openBracketCount === 0) {
                            closingBracket = j;
                            break;
                        }
                    }
                }
                result.push(content.slice(i, closingBracket + 1));
                i = closingBracket;
            } else if (token.value === '[') {
                let closingBracket = -1;
                let openBracketCount = 0;

                for (let j = i; j < content.length; j++) {
                    if (content[j].value === '[') {
                        openBracketCount++;
                    } else if (content[j].value === ']') {
                        openBracketCount--;
                        if (openBracketCount === 0) {
                            closingBracket = j;
                            break;
                        }
                    }
                }
                result.push(content.slice(i, closingBracket + 1));
                // console.log('content', content);
                // console.log('slice', content.slice(i, closingBracket + 1));
                i = closingBracket;
            } else {
                result.push([token]);
            }
        }

        return result;
    }

    divideExpressionList(content: Tokenized[]): Tokenized[][] {
        const result: Tokenized[][] = [];
        const ifComma = content.some((item) => item.value === ',');
        if (!ifComma && content.length) {
            result.push(content);
        } else {
            for (let i = 0; i < content.length; i++) {
                const commaIndex = content.findIndex(
                    (item, index) => item.value === ',' && index >= i
                );

                if (commaIndex !== -1) {
                    result.push(content.slice(i, commaIndex));
                    result.push([content[commaIndex]]);
                    i = commaIndex;
                } else {
                    if (i !== content.length) {
                        result.push(content.slice(i, content.length));
                        break;
                    }
                }
            }
        }

        return result;
    }

    shuffleTerms(content: Tokenized[][]): Tokenized[][] {
        const operandIndexes = content
            .map((item, index) => {
                if (
                    operands.includes(item[0].value as string) ||
                    unaryOperands.includes(item[0].value as string)
                ) {
                    return index;
                }
            })
            .filter((item) => item !== undefined);
        if (operandIndexes.length !== 0) {
            operandIndexes.forEach((index) => {
                const operand = content[index!];
                const right = content[index! + 1];
                content[index!] = right;
                content[index! + 1] = operand;
            });
        }
        return content;
    }
}
